src/Microsoft.NetCore.Analyzers/Core/Security/DoNotCallDangerousMethodsInDeserialization.cs

```diff
@@ -0,0 +1,219 @@
+                            {
+                                if (member.Kind == SymbolKind.Method)
+                                {
+                                    var methodSymbol = member as IMethodSymbol;
+
+                                    if (methodSymbol.GetAttributes().Any(s => attributeTypeSymbols.Contains(s.AttributeClass))
+                                        || methodSymbol.MetadataName == "OnDeserialization")
+                                    {
+                                        var parameters = methodSymbol.GetParameters();
+
+                                        if (parameters == null
+                                            || parameters.Length != 1
+                                            || !parameters[0].Type.Equals(streamingContextTypeSymbol))
+                                        {
+                                            return;
+                                        }
+
+                                        var visitedMethodSymbols = new List<IMethodSymbol>();
+                                        var invocationOperations = FindDangerousMethodInvocatioinOperation(methodSymbol, compilation, dangerousMethodSymbols, ref visitedMethodSymbols);
+
+                                        foreach (var item in invocationOperations)
+                                        {
+                                            symbolAnalysisContext.ReportDiagnostic(
+                                                item.CreateDiagnostic(
+                                                    Rule,
+                                                    classSymbol.MetadataName,
+                                                    methodSymbol.MetadataName,
+                                                    item.TargetMethod.MetadataName));
+                                        }
+                                    }
+                                }
+                            }
+                        },
+                        SymbolKind.NamedType);
+                });
+
+            /// <summary>
+            /// Traverse every invocation in the method to find all the dangerous function it invokes.
+            /// </summary>
+            /// <param name="methodSymbol">The Symbol of the method which is going to be analyzed</param>
+            /// <param name="compilation">Compilation context</param>
+            /// <param name="dangerousMethodSymbols">Symbols reprensent dangerous methods</param>
+            /// <param name="visitedMethodSymbols">Methods that have been analyzed</param>
+            IEnumerable<IInvocationOperation> FindDangerousMethodInvocatioinOperation(
+                IMethodSymbol methodSymbol,
+                Compilation compilation,
+                ImmutableArray<IMethodSymbol> dangerousMethodSymbols,
+                ref List<IMethodSymbol> visitedMethodSymbols)
+            {
+                var result = new List<IInvocationOperation>();
```

Comment:
Nit: Could pass this `result` List as an input parameter, and just have FindDangerousMethodInvocationOperations populate it.  That would avoid unnecessary List allocations and copying during recursion.